app/services/course/assessment/programming_evaluation_service.rb

```diff
@@ -90,37 +93,39 @@ def initialize(course, language, memory_limit, # rubocop:disable Metrics/Paramet
     @timeout = timeout || DEFAULT_TIMEOUT
   end
 
-  # Creates the evaluation, waits for its completion, then returns the result.
+  # Evaluate the package in a Docker container and return the output that matters.
   #
   # @return [Result]
+  # @raise [Timeout::Error] When the evaluation timeout has elapsed.
   def execute
-    evaluation = create_evaluation
-    wait_for_evaluation(evaluation)
-    Result.new(evaluation.stdout, evaluation.stderr, evaluation.test_report,
-               evaluation.exit_code, evaluation.id)
-  ensure
-    evaluation.destroy! if evaluation
+    stdout, stderr, test_report, exit_code = Timeout.timeout(@timeout) { evaluate_in_container }
```

Comment:
Right now it's not set by the user. This is the evaluation timeout (default 5 minutes).

The user set one is `time_limit`.